### DISCUSSION TITLE
Is possible to support JQ syntax language?

### DISCUSSION THREAD

**Discussion by aco1731** [grafana/grafana-infinity-datasource#441] [Ideas]
Is possible to support jq tool syntax language to transform the JSON? 

The only options i saw was UQL and JSONata. 

Or is it possible to do anything with JSONata that is possible with jq syntax language?

**Comment by yesoreyeram** [comment]
Yes it's is already possible via JSONata. Can you post sample data and what you want to achieve?

**Reply by aco1731** [reply]
GRAFANA VERSION: 8.3.4
INFINITY PLUGIN: 0.8.8

I received a array of:

 {
      "kind": "tm:ltm:virtual:virtualstate",
      "name": "vs_BLABLA",
      "partition": "Common",
      "fullPath": "/Common/vs_BLABLA",
      "generation": 1,
      "selfLink": "https://localhost/mgmt/tm/ltm/virtual/~Common~vs_BLABLA?ver=15.1.5",
      "addressStatus": "yes",
      "autoLasthop": "default",
      "cmpEnabled": "yes",
      "connectionLimit": 0,
      "creationTime": "2022-07-27T09:40:13Z",
      "description": "123 teste",
      "destination": "/Common/192.168.1.1:80",
      "enabled": true,
      "evictionProtected": "disabled",
      "gtmScore": 0,
      "ipProtocol": "any",
      "lastModifiedTime": "2022-07-27T09:40:13Z",
      "mask": "any",
      "mirror": "disabled",
      "mobileAppTunnel": "disabled",
      "nat64": "disabled",
      "rateLimit": "disabled",
      "rateLimitDstMask": 0,
      "rateLimitMode": "object",
      "rateLimitSrcMask": 0,
      "serversslUseSni": "disabled",
      "serviceDownImmediateAction": "none",
      "source": "0.0.0.0/0",
      "sourceAddressTranslation": {
        "type": "none"
      }

I just want to create a variable that unified name and destination like:  vs_BLABLA ( /Common/192.168.1.1:80 ).

It would be good to also be able to regex just the IP part: vs_BLABLA (192.168.1.1).

I can easily do it with JQ.. But how do i enable JQ/UQL  using variables with Infinity?

**Comment by aco1731** [comment]
I would like to transform the source JSON below to variable in dashboard that allow me to search for pool or IP address.

Using jq, i was able to do that kind of manipulation/transform: 

```
curl -sk -u admin:pass   -H "Content-Type: application/json" -X GET https://192.168.1.100/mgmt/tm/ltm/virtual | jq '[.items[] | select(.pool !=  null) | .pool + "  (" + (.destination|capture("(?<ip>[[:digit:].]+)").ip) + ")"   ]'
[
  "/Common/pool_HAPL-8281  (10.70.184.193)",
  "/Common/pool_APL-8009  (10.31.220.8)",
  "/Common/pool_TESTE  (192.168.1.200)",
  "/Common/pool_NTP  (192.168.1.210)",
  "/Common/pool_WWW  (10.138.31.2)",

])
```

How can i do something similar with Infinity? I would like to use a langague that i already know like jq... If not possible what path should i take? JSONPath/JSONata/UQL? Im a bit lost.


source JSON:
```json
{
"items": [
    {
      "kind": "tm:ltm:virtual:virtualstate",
      "name": "vs_TESTE-VIP",
      "partition": "Common",
      "fullPath": "/Common/vs_TESTE-VIP",
      "generation": 1,
      "selfLink": "https://localhost/mgmt/tm/ltm/virtual/~Common~vs_TESTE-VIP?ver=15.1.5",
      "addressStatus": "yes",
      "autoLasthop": "disabled",
      "cmpEnabled": "yes",
      "connectionLimit": 0,
      "creationTime": "2022-07-15T05:04:26Z",
      "description": "VIP DE TESTE",
      "destination": "/Common/192.168.1.200:80",
      "enabled": true,
      "evictionProtected": "disabled",
      "gtmScore": 0,
      "ipProtocol": "tcp",
      "lastModifiedTime": "2022-07-27T06:09:57Z",
      "mask": "255.255.255.255",
      "mirror": "disabled",
      "mobileAppTunnel": "disabled",
      "nat64": "disabled",
      "pool": "/Common/pool_TESTE-FARM",
      "poolReference": {
        "link": "https://localhost/mgmt/tm/ltm/pool/~Common~pool_TESTE-FARM?ver=15.1.5"
      },
      "rateLimit": "disabled",
      "rateLimitDstMask": 0,
      "rateLimitMode": "object",
      "rateLimitSrcMask": 0,
      "serversslUseSni": "disabled",
      "serviceDownImmediateAction": "none",
      "source": "0.0.0.0/0",
      "sourceAddressTranslation": {
        "type": "automap"
      },
      "sourcePort": "preserve",
      "synCookieStatus": "not-activated",
      "translateAddress": "enabled",
      "translatePort": "enabled",
      "vlansDisabled": true,
      "vsIndex": 7,
      "policiesReference": {
        "link": "https://localhost/mgmt/tm/ltm/virtual/~Common~vs_TESTE-VIP/policies?ver=15.1.5",
        "isSubcollection": true
      },
      "profilesReference": {
        "link": "https://localhost/mgmt/tm/ltm/virtual/~Common~vs_TESTE-VIP/profiles?ver=15.1.5",
        "isSubcollection": true
      }
    },
    {
      "kind": "tm:ltm:virtual:virtualstate",
      "name": "vs_TESTEPAPLCCF-ANY",
      "partition": "Common",
      "fullPath": "/Common/vs_TESTEPAPLCCF-ANY",
      "generation": 1,
      "selfLink": "https://localhost/mgmt/tm/ltm/virtual/~Common~vs_TESTEPAPLCCF-ANY?ver=15.1.5",
      "addressStatus": "yes",
      "autoLasthop": "default",
      "cmpEnabled": "yes",
      "connectionLimit": 0,
      "creationTime": "2022-07-26T13:28:34Z",
      "destination": "/Common/10.31.220.9:0",
      "enabled": true,
      "evictionProtected": "disabled",
      "gtmScore": 0,
      "ipProtocol": "tcp",
      "lastModifiedTime": "2022-07-26T13:28:34Z",
      "mask": "255.255.255.255",
      "mirror": "disabled",
      "mobileAppTunnel": "disabled",
      "nat64": "disabled",
      "rateLimit": "disabled",
      "rateLimitDstMask": 0,
      "rateLimitMode": "object",
      "rateLimitSrcMask": 0,
      "serversslUseSni": "disabled",
      "serviceDownImmediateAction": "none",
      "source": "0.0.0.0/0",
      "sourceAddressTranslation": {
        "type": "none"
      },
      "sourcePort": "preserve",
      "synCookieStatus": "not-activated",
      "translateAddress": "enabled",
      "translatePort": "disabled",
      "vlansDisabled": true,
      "vsIndex": 8,
      "policiesReference": {
        "link": "https://localhost/mgmt/tm/ltm/virtual/~Common~vs_TESTEPAPLCCF-ANY/policies?ver=15.1.5",
        "isSubcollection": true
      },
      "profilesReference": {
        "link": "https://localhost/mgmt/tm/ltm/virtual/~Common~vs_TESTEPAPLCCF-ANY/profiles?ver=15.1.5",
        "isSubcollection": true
      }
    }
  ]
}
```


Thank you in advance...

**Comment by aco1731** [comment]
curl -sk -u admin:pass   -H "Content-Type: application/json" -X GET https://192.168.1.100/mgmt/tm/ltm/virtual | jq '[.items[] | select(.pool !=  null) | .pool + "  (" + (.destination|capture("(?<ip>[[:digit:].]+)").ip) + ")"   ]'
[
  "/Common/pool_HAPL-8281  (10.70.184.193)",
  "/Common/pool_APL-8009  (10.31.220.8)",
  "/Common/pool_TESTE  (192.168.1.200)",
  "/Common/pool_NTP  (192.168.1.210)",
  "/Common/pool_WWW  (10.138.31.2)",

])


How can i obtain something similar using UQL?

**Comment by rapcore2** [comment]
@yesoreyeram Any example how to use jq tool syntax language ?

**Comment by yesoreyeram** [comment]
`jq` support is coming in 3.0.0 via https://github.com/grafana/grafana-infinity-datasource/pull/1143
